### ISSUE TITLE
Unnecessary rebuilding of targets during build after failed build

### ISSUE THREAD

**Issue by bitc** [ndmitchell/shake#250]
This happens from time to time, but is hard to reproduce.

A build will fail, but after successfully building some intermediate targets. When the build is run again, the intermediate targets that were previously successfully built are unnecessarily rebuilt.

Example:

```
$ ./build.sh 
# g++ (for _build/GoodFile.o)
# g++ (for _build/BadFile.o)
src/BadFile.cpp: some compiler error
Error when running Shake build system:
* _build/myprogram
* _build/BadFile.o
user error (Development.Shake.cmd, system command failed
Command: g++ -c src/BadFile.cpp -o _build/BadFile.o -MMD -MF _build/BadFile.m -Isrc
Exit code: 1
Stderr:
src/BadFile.cpp: some compiler error

$ ./build.sh 
# g++ (for _build/GoodFile.o)
# g++ (for _build/BadFile.o)
src/BadFile.cpp: some compiler error
Error when running Shake build system:
* _build/myprogram
* _build/BadFile.o
user error (Development.Shake.cmd, system command failed
Command: g++ -c src/BadFile.cpp -o _build/BadFile.o -MMD -MF _build/BadFile.m -Isrc
Exit code: 1
Stderr:
src/BadFile.cpp: some compiler error
```

Notice that during the second build attempt, `GoodFile.cpp` is being compiled again for no reason. The Build rules I'm using for C++ are pretty much identical to the ones in the Shake manual.

I can try to create a reproducible test case, but I figured I'd post this in the meantime to see if anyone has any feedback

**Comment by ndmitchell** [comment]
Could this be a race condition? You start compiling GoodFile.cpp and BadFile.cpp simultaneously. If GoodFile.cpp completes before BadFile.cpp raises an error then it will be saved. If BadFile finishes first (which, if it has an error, is likely) then GoodFile will be killed or will finish after Shake has shut down.

**Comment by bitc** [comment]
I don't think this is the problem. I'll try to see if I can make a small reproducible example

**Comment by bitc** [comment]
Here is a test case that shows the bug. 

I'm using the shake build script from the beginning of the [Shake Manual](http://shakebuild.com/manual) (the one that compiles a C program).

Make sure to compile the shake build script to the executable `./Build`, and then try this bash script:

``` bash
#!/bin/bash

set -x

# Make sure to start from a clean state
rm -f main.c f1.c f2.c f3.c f4.c f5.c f6.c incl.h
rm -rf _build


# Generate a header file

cat > incl.h << EOF
int foo();
EOF

# Generate some source files

cat > f1.c << EOF
#include "incl.h"
EOF

cp f1.c f2.c
cp f1.c f3.c
cp f1.c f4.c
cp f1.c f5.c
cp f1.c f6.c

cat > main.c << EOF
int main() { return 0; }
EOF


# Perform a build (this should succeed)
./Build

# Modify f3.c, introducing an error
echo "int blah()}" >> f3.c

# Modify the rest of the source files
touch incl.h f1.c f2.c f4.c f5.c f6.c

# Perform a build (this should succesfully build some object files and then
# fail on f3.o)
./Build

# Perform another build. This should try building only f3.o (which previously
# failed). Instead it rebuilds all the object files that it already succesfully
# built in the previous step
./Build
```

Here is the output when I run it:

```
+ rm -f main.c f1.c f2.c f3.c f4.c f5.c f6.c incl.h
+ rm -rf _build
+ cat
+ cat
+ cp f1.c f2.c
+ cp f1.c f3.c
+ cp f1.c f4.c
+ cp f1.c f5.c
+ cp f1.c f6.c
+ cat
+ ./Build
# gcc (for _build/main.o)
# gcc (for _build/f6.o)
# gcc (for _build/f5.o)
# gcc (for _build/f4.o)
# gcc (for _build/f3.o)
# gcc (for _build/f2.o)
# gcc (for _build/f1.o)
# gcc (for _build/run)
Build completed in 0:01m
+ echo 'int blah()}'
+ touch incl.h f1.c f2.c f4.c f5.c f6.c
+ ./Build
# gcc (for _build/f6.o)
# gcc (for _build/f5.o)
# gcc (for _build/f4.o)
# gcc (for _build/f3.o)
f3.c: In function 'blah':
f3.c:2:11: error: expected declaration specifiers before '}' token
 int blah()}
           ^
f3.c:2:11: error: expected '{' at end of input
Error when running Shake build system:
* _build/run
* _build/f3.o
user error (Development.Shake.cmd, system command failed
Command: gcc -c f3.c -o _build/f3.o -MMD -MF _build/f3.m
Exit code: 1
Stderr:
f3.c: In function 'blah':
f3.c:2:11: error: expected declaration specifiers before '}' token
 int blah()}
           ^
f3.c:2:11: error: expected '{' at end of input
)

+ ./Build
# gcc (for _build/f6.o)
# gcc (for _build/f5.o)
# gcc (for _build/f4.o)
# gcc (for _build/f3.o)
f3.c: In function 'blah':
f3.c:2:11: error: expected declaration specifiers before '}' token
 int blah()}
           ^
f3.c:2:11: error: expected '{' at end of input
Error when running Shake build system:
* _build/run
* _build/f3.o
user error (Development.Shake.cmd, system command failed
Command: gcc -c f3.c -o _build/f3.o -MMD -MF _build/f3.m
Exit code: 1
Stderr:
f3.c: In function 'blah':
f3.c:2:11: error: expected declaration specifiers before '}' token
 int blah()}
           ^
f3.c:2:11: error: expected '{' at end of input
)
```

Notice at the very end how it re-builds `f6.o` `f5.o` and `f4.o`, even though they were already succesfully built before and nothing has changed on the filesystem.

I'm using shake 0.15.2 installed from hackage with cabal-install (GHC 7.6.3)

**Comment by ndmitchell** [comment]
Thanks for the detailed test case. With that, I was able to reproduce, and it was indeed unexpected. Investigating further, I now understand why it happens, and it is actually by design.

After the `touch`, all files are dirty, that includes `f6.c`, `f6.o`, `f3.c`, `f3.o` and `incl.h`. Shake proceeds by noticing `f6.o` is dirty, and recompiling it with `gcc`. It then encounters the line `needMakefileDependencies` which boils down to `need ["f6.c","incl.h"]`. Since those files are dirty, Shake puts actions to make them into the pool, and continues with `f3.o` (or in your case, `f5.o`, then `f4.o`). When `f3.o` throws an exception it immediately aborts, without running `incl.h`.

As a temporary workaround, if you had passed `--keep-going` (staunch mode) then it would have cleared the queues, and then only `f3.o` would have failed.

Longer term, I can think of two things:

1. There is probably an advantage to prioritising items that are dependencies of rules that have already started running. In some way, when `f6.o` finishes it might be able to "delegate" it's time slice to `f6.c`. But at the same time, the link rule is already running then, so in some ways `f3.o` is just as valid to switch to.
2. If a rule is "super-cheap" (like a need on a file that does not have a meaningful rule, and is just a get modification time) then maybe I should run it without spawning out to the pool, like I already do when checking it. The problem here is that while file rules are super cheap, the pattern matching can be arbitrarily expensive, and running expensive things outside the pool is a bad idea.

So I wonder if this is a "surprising, but won't fix"?

**Comment by bitc** [comment]
Thanks for the reply! After reading it several times I finally understand your explanation.

I think I may have another idea:

3) After a rule fails (like `f3.o`), instead of immediately exiting: have shake first go through all the remaining rules in the pool, looking for "super-cheap" ones and marking them as built as necessary. This is kind of like `--keep-going`, but after the error it will skip any rules that are not "super-cheap". (A new command line flag could be added to shake to enable this mode if it adds too much overhead by default [you mentioned expensive pattern matching], although i personally think that this would be good default behavior and the way that most users would expect shake to behave).

**Comment by ndmitchell** [comment]
Yep, that's plausible, and not such a big change under normal circumstances. I'd still need some way to mark rules as super-cheap. I agree that it would be nice if it "just worked" here.

**Comment by bitc** [comment]
Is there an update regarding this issue? (I can't imagine that I'm the only one affected by it). Thanks

**Comment by bitc** [comment]
Bump. Is there a plan to tackle this issue?
I'm back to coding a large C++ project with shake and this issue causes lots of pain.
I could try to implement the proposed "super-cheap" technique, but the shake internals might be too advanced for me.
